File: app/src/main/java/com/example/customappactionsheet/ActionSheetAdapter.java
package com.example.customappactionsheet;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

public class ActionSheetAdapter extends RecyclerView.Adapter<BaseActionSheetViewHolder> {

    private static final String TAG = ActionSheetAdapter.class.getSimpleName();
    private static ActionSheetListener listener;
    private static final int TopView = 0;
    private static final int CenterView = 1;
    private static final int BottomView = 2;
    private static final int DefaultView = 4;
    private static List<ActionSheetViewHolderModel> list = new ArrayList<ActionSheetViewHolderModel>();

    public ActionSheetAdapter(ActionSheetListener listener) {
        this.listener = listener;
    }

    @NonNull
    @Override
    public BaseActionSheetViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
        View view;
        switch (viewType) {
            case TopView:
                Log.d(TAG,"TopView");
                view = layoutInflater.inflate(R.layout.action_sheet_top_cell, parent, false);
                return new ActionSheetTopViewHolder(parent.getContext(), listener, view);
            case CenterView:
                Log.d(TAG,"CenterView");
                view = layoutInflater.inflate(R.layout.action_sheet_center_cell, parent, false);
                return new ActionSheetCenterViewHolder(parent.getContext(), listener, view);
            case BottomView:
                Log.d(TAG,"BottomView");
                view = layoutInflater.inflate(R.layout.action_sheet_bottom_cell, parent, false);
                return new ActionSheetBottomViewHolder(parent.getContext(), listener, view);
            default:
                Log.d(TAG,"default");
                view = layoutInflater.inflate(R.layout.action_sheet_cell, parent, false);
                return new ActionSheetViewHolder(parent.getContext(), listener, view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull BaseActionSheetViewHolder holder, int position) {
        holder.bindDataToViewHolder(list.get(position), position);
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (list.size() == 1 && position == 0) {
            return DefaultView;
        } else if (list.size() > 1 && position == 0) {
            return TopView;
        } else if (list.size() > 1 && position + 1 == list.size()) {
            return BottomView;
        } else {
            return CenterView;
        }
    }

    public void setItems(List<ActionSheetViewHolderModel> list) {
        this.list.clear();
        this.list.addAll(list);
        notifyDataSetChanged();
    }
}

File: app/src/main/java/com/example/customappactionsheet/ActionSheetAdapter.kt
package com.example.customappactionsheet

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

class ActionSheetAdapter : RecyclerView.Adapter<BaseActionSheetViewHolder> {

    companion object {
        private val TAG : String = ActionSheetAdapter::class.java.simpleName
        private lateinit var listener : ActionSheetListener
        private const val TopView : Int = 0
        private const val CenterView : Int = 1
        private const val BottomView : Int = 2
        private const val DefaultView : Int = 4
        private val list : MutableList<ActionSheetViewHolderModel> = mutableListOf<ActionSheetViewHolderModel>()
    }

    constructor(listener: ActionSheetListener) {
        ActionSheetAdapter.listener = listener
    }

    override fun onCreateViewHolder(parent : ViewGroup, viewType : Int) : BaseActionSheetViewHolder {
        val layoutInflater : LayoutInflater = LayoutInflater.from(parent.context)
        val view : View
        return when(viewType) {
            TopView -> { Log.d(TAG,"TopView ActionSheetTopViewHolder")
                view = layoutInflater.inflate(R.layout.action_sheet_top_cell, parent, false)
                return ActionSheetTopViewHolder(listener, view)
            }
            CenterView -> { Log.d(TAG,"CenterView ActionSheetCenterViewHolder")
                view = layoutInflater.inflate(R.layout.action_sheet_center_cell, parent, false)
                return ActionSheetCenterViewHolder(listener, view)
            }
            BottomView -> { Log.d(TAG,"BottomView ActionSheetBottomViewHolder")
                view = layoutInflater.inflate(R.layout.action_sheet_bottom_cell, parent, false)
                return ActionSheetBottomViewHolder(listener, view)
            }
            else -> { Log.d(TAG,"else ActionSheetViewHolder")
                view = layoutInflater.inflate(R.layout.action_sheet_cell, parent, false)
                return ActionSheetViewHolder(listener, view)
            }
        }
    }

    override fun onBindViewHolder(holder : BaseActionSheetViewHolder, position : Int) {
        holder.bindDataToViewHolder(list.get(position), position)
    }

    override fun getItemCount() : Int {
        return list.size
    }

    override fun getItemViewType(position : Int) : Int {
        if (list.size == 1 && position == 0) {
            return DefaultView
        } else if (list.size > 1 && position == 0) {
            return TopView
        } else if (list.size > 1 && position + 1 == list.size) {
            return BottomView
        } else {
            return CenterView
        }
    }

    public fun setItems(items : List<ActionSheetViewHolderModel>) {
        list.clear()
        list.addAll(items)
        notifyDataSetChanged()
    }
}